authz_analyzer/datastores/aws/aws_ptrp_package/aws_ptrp/services/s3/s3_actions.py

```diff
@@ -60,26 +63,76 @@ def add(self, actions: Set[S3Action], stmt_relative_id_objects_regex: Optional[s
+        for s_regex in self.stmt_relative_id_objects_regexes:
+            if (
+                any(is_aws_regex_full_subset(o_regex, s_regex) for o_regex in other.stmt_relative_id_objects_regexes)
+                is False
+            ):
+                return False
+        return True
+
+    def difference(self, other: ResolvedActionsSingleStmt):
         """
-        the function checks that each object regex in self is a 'full subset' of any regex in other, and if so, it removes the object actions (in addition to the bucket actions)
+        This functions differs the s3 actions between 'self' and 'other'
+        Bucket actions will always be removed. In case of object actions, it is depends on the relative object regexes in each one
+        Generally, if each object regex in self is a 'full subset' of at least one regex in other, we removes the object actions (in addition to the bucket actions)
+        If some(or both) of the statements are annotated with NotResource, in some cases we can solve the problem with the complements of each group
         """
         difference_also_on_object_actions = False
         if isinstance(other, ResolvedS3BucketActions):
-            if other.stmt_relative_id_objects_regexes:
-                for s_regex in self.stmt_relative_id_objects_regexes:
-                    difference_also_on_object_actions = any(
-                        is_aws_regex_full_subset(o_regex, s_regex) for o_regex in other.stmt_relative_id_objects_regexes
+            self_not_resource_annotated = self.is_not_resource_annotated()
+            other_not_resource_annotated = other.is_not_resource_annotated()
+            if self_not_resource_annotated is False and other_not_resource_annotated is False:
+                if other.stmt_relative_id_objects_regexes:
+                    difference_also_on_object_actions = self.are_all_object_regexes_full_subset_of_any_regex_in_other(
+                        other
+                    )
+
+            elif self_not_resource_annotated is True and other_not_resource_annotated is True:
+                # S < O <-> O' < S' (S' = complement(S), O' = complement(O)
+                # If both statements are annotated with NotResource, we can solve the problem with the complements of each group
+                if self.stmt_relative_id_objects_regexes:
+                    difference_also_on_object_actions = other.are_all_object_regexes_full_subset_of_any_regex_in_other(
+                        self
                     )
 
-                    if difference_also_on_object_actions is False:
-                        break
+            elif self_not_resource_annotated is True and other_not_resource_annotated is False:
+                # We cant really solve the problem with complements, so we will assume that we can remove the object actions
+                # Only if other.stmt_relative_id_objects_regexes as a wildcard
+                difference_also_on_object_actions = any(
```

Comment:
Do you think you need this optimization?  didn't you handle this in the resolve function ?